// molstar/lib/commonjs/mol-gl/renderable/mesh.js
"use strict";
/**
 * Copyright (c) 2018-2025 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeshSchema = void 0;
exports.MeshRenderable = MeshRenderable;
const renderable_1 = require("../renderable.js");
const render_item_1 = require("../webgl/render-item.js");
const schema_1 = require("./schema.js");
const shader_code_1 = require("../shader-code.js");
const mol_util_1 = require("../../mol-util/index.js");
exports.MeshSchema = {
    ...schema_1.BaseSchema,
    aGroup: (0, schema_1.AttributeSpec)('float32', 1, 0),
    aPosition: (0, schema_1.AttributeSpec)('float32', 3, 0),
    aNormal: (0, schema_1.AttributeSpec)('float32', 3, 0),
    elements: (0, schema_1.ElementsSpec)('uint32'),
    dVaryingGroup: (0, schema_1.DefineSpec)('boolean'),
    dFlatShaded: (0, schema_1.DefineSpec)('boolean'),
    uDoubleSided: (0, schema_1.UniformSpec)('b', 'material'),
    dFlipSided: (0, schema_1.DefineSpec)('boolean'),
    dIgnoreLight: (0, schema_1.DefineSpec)('boolean'),
    dCelShaded: (0, schema_1.DefineSpec)('boolean'),
    dXrayShaded: (0, schema_1.DefineSpec)('string', ['off', 'on', 'inverted']),
    dTransparentBackfaces: (0, schema_1.DefineSpec)('string', ['off', 'on', 'opaque']),
    uBumpFrequency: (0, schema_1.UniformSpec)('f', 'material'),
    uBumpAmplitude: (0, schema_1.UniformSpec)('f', 'material'),
    meta: (0, schema_1.ValueSpec)('unknown')
};
function MeshRenderable(ctx, id, values, state, materialId, transparency, globals) {
    const schema = { ...schema_1.GlobalUniformSchema, ...schema_1.GlobalTextureSchema, ...schema_1.GlobalDefineSchema, ...schema_1.InternalSchema, ...exports.MeshSchema };
    const renderValues = {
        ...values,
        uObjectId: mol_util_1.ValueCell.create(id),
        dLightCount: mol_util_1.ValueCell.create(globals.dLightCount),
        dColorMarker: mol_util_1.ValueCell.create(globals.dColorMarker),
    };
    const shaderCode = shader_code_1.MeshShaderCode;
    const renderItem = (0, render_item_1.createGraphicsRenderItem)(ctx, 'triangles', shaderCode, schema, renderValues, materialId, transparency);
    return (0, renderable_1.createRenderable)(renderItem, renderValues, state);
}
