// molstar/lib/commonjs/mol-gl/renderable/spheres.js
"use strict";
/**
 * Copyright (c) 2019-2025 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpheresSchema = void 0;
exports.SpheresRenderable = SpheresRenderable;
const renderable_1 = require("../renderable.js");
const render_item_1 = require("../webgl/render-item.js");
const schema_1 = require("./schema.js");
const shader_code_1 = require("../shader-code.js");
const mol_util_1 = require("../../mol-util/index.js");
exports.SpheresSchema = {
    ...schema_1.BaseSchema,
    ...schema_1.SizeSchema,
    uTexDim: (0, schema_1.UniformSpec)('v2'),
    tPositionGroup: (0, schema_1.TextureSpec)('image-float32', 'rgba', 'float', 'nearest'),
    padding: (0, schema_1.ValueSpec)('number'),
    uDoubleSided: (0, schema_1.UniformSpec)('b', 'material'),
    dIgnoreLight: (0, schema_1.DefineSpec)('boolean'),
    dCelShaded: (0, schema_1.DefineSpec)('boolean'),
    dXrayShaded: (0, schema_1.DefineSpec)('string', ['off', 'on', 'inverted']),
    dTransparentBackfaces: (0, schema_1.DefineSpec)('string', ['off', 'on', 'opaque']),
    dSolidInterior: (0, schema_1.DefineSpec)('boolean'),
    dClipPrimitive: (0, schema_1.DefineSpec)('boolean'),
    dApproximate: (0, schema_1.DefineSpec)('boolean'),
    uAlphaThickness: (0, schema_1.UniformSpec)('f'),
    uBumpFrequency: (0, schema_1.UniformSpec)('f', 'material'),
    uBumpAmplitude: (0, schema_1.UniformSpec)('f', 'material'),
    lodLevels: (0, schema_1.ValueSpec)('unknown'),
    centerBuffer: (0, schema_1.ValueSpec)('float32'),
    groupBuffer: (0, schema_1.ValueSpec)('float32'),
};
function SpheresRenderable(ctx, id, values, state, materialId, transparency, globals) {
    const schema = { ...schema_1.GlobalUniformSchema, ...schema_1.GlobalTextureSchema, ...schema_1.GlobalDefineSchema, ...schema_1.InternalSchema, ...exports.SpheresSchema };
    const renderValues = {
        ...values,
        uObjectId: mol_util_1.ValueCell.create(id),
        dLightCount: mol_util_1.ValueCell.create(globals.dLightCount),
        dColorMarker: mol_util_1.ValueCell.create(globals.dColorMarker),
    };
    const shaderCode = shader_code_1.SpheresShaderCode;
    const renderItem = (0, render_item_1.createGraphicsRenderItem)(ctx, 'triangles', shaderCode, schema, renderValues, materialId, transparency);
    return (0, renderable_1.createRenderable)(renderItem, renderValues, state);
}
