// molstar/lib/commonjs/extensions/mvs/tree/mvs/mvs-builder.js
"use strict";
/**
 * Copyright (c) 2023-2025 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Primitives = exports.VolumeRepresentation = exports.Volume = exports.Representation = exports.Component = exports.Structure = exports.Parse = exports.Download = exports.Animation = exports.Root = void 0;
exports.createMVSBuilder = createMVSBuilder;
exports.builderDemo = builderDemo;
const object_1 = require("../../../../mol-util/object.js");
const mvs_data_1 = require("../../mvs-data.js");
/** Create a new MolViewSpec builder containing only a root node. Example of MVS builder usage:
 *
 * ```
 * const builder = createMVSBuilder();
 * builder.canvas({ background_color: 'white' });
 * const struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
 * struct.component().representation().color({ color: '#3050F8' });
 * console.log(JSON.stringify(builder.getState()));
 * ```
 */
function createMVSBuilder(params = {}) {
    return new Root(params);
}
/** Base class for MVS builder pointing to anything */
class _Base {
    constructor(_root, _node) {
        this._root = _root;
        this._node = _node;
    }
    /** Create a new node, append as child to current _node, and return the new node */
    addChild(kind, params_) {
        var _a;
        var _b;
        const { params, custom, ref } = splitParams(params_);
        const node = {
            kind,
            params,
            custom,
            ref,
        };
        (_a = (_b = this._node).children) !== null && _a !== void 0 ? _a : (_b.children = []);
        this._node.children.push(node);
        return node;
    }
}
/** MVS builder pointing to the 'root' node */
class Root extends _Base {
    constructor(params_) {
        const { custom, ref } = params_;
        const node = { kind: 'root', custom, ref };
        super(undefined, node);
        this._animation = undefined;
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
        this.primitives = bindMethod(this, PrimitivesMixinImpl, 'primitives');
        this.primitivesFromUri = bindMethod(this, PrimitivesMixinImpl, 'primitivesFromUri');
        this._root = this;
    }
    /** Return the current state of the builder as object in MVS format. */
    getState(metadata) {
        return {
            root: (0, object_1.deepClone)(this._node),
            metadata: mvs_data_1.GlobalMetadata.create(metadata),
        };
    }
    // omitting `saveState`, filesystem operations are responsibility of the caller code (platform-dependent)
    /** Return the current state of the builder as a snapshot object to be used in multi-state . */
    getSnapshot(metadata) {
        return {
            root: (0, object_1.deepClone)(this._node),
            metadata: { ...metadata },
            animation: (this === null || this === void 0 ? void 0 : this._animation) ? (0, object_1.deepClone)(this._animation.node) : undefined,
        };
    }
    /** Add a 'camera' node and return builder pointing to the root. 'camera' node instructs to set the camera position and orientation. */
    camera(params) {
        this.addChild('camera', params);
        return this;
    }
    /** Add a 'canvas' node and return builder pointing to the root. 'canvas' node sets canvas properties. */
    canvas(params) {
        this.addChild('canvas', params);
        return this;
    }
    /** Add a 'download' node and return builder pointing to it. 'download' node instructs to retrieve a data resource. */
    download(params) {
        return new Download(this._root, this.addChild('download', params));
    }
    animation(params = {}) {
        var _a;
        (_a = this._animation) !== null && _a !== void 0 ? _a : (this._animation = new Animation(params));
        return this._animation;
    }
    /** Modifies custom state of the root */
    extendRootCustomState(custom) {
        this._node.custom = { ...this._node.custom, ...custom };
        return this;
    }
}
exports.Root = Root;
class Animation {
    constructor(parameters) {
        this._node = {
            kind: 'animation',
            children: [],
            ...splitParams(parameters),
        };
    }
    get node() {
        return this._node;
    }
    interpolate(params) {
        const node = {
            kind: 'interpolate',
            ...splitParams(params)
        };
        this._node.children.push(node);
        return this;
    }
}
exports.Animation = Animation;
/** MVS builder pointing to a 'download' node */
class Download extends _Base {
    /** Add a 'parse' node and return builder pointing to it. 'parse' node instructs to parse a data resource. */
    parse(params) {
        return new Parse(this._root, this.addChild('parse', params));
    }
}
exports.Download = Download;
/** Subsets of 'structure' node params which will be passed to individual builder functions. */
const StructureParamsSubsets = {
    model: ['block_header', 'block_index', 'model_index', 'coordinates_ref'],
    assembly: ['block_header', 'block_index', 'model_index', 'assembly_id', 'coordinates_ref'],
    symmetry: ['block_header', 'block_index', 'model_index', 'ijk_min', 'ijk_max', 'coordinates_ref'],
    symmetry_mates: ['block_header', 'block_index', 'model_index', 'radius', 'coordinates_ref'],
};
/** MVS builder pointing to a 'parse' node */
class Parse extends _Base {
    /** Add a 'structure' node representing a "model structure", i.e. includes all coordinates from the original model without applying any transformations.
     * Return builder pointing to the new node. */
    modelStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'model',
            ...(0, object_1.pickObjectKeys)(params, [...StructureParamsSubsets.model]),
            custom: params.custom,
            ref: params.ref,
        }));
    }
    /** Add a 'structure' node representing an "assembly structure", i.e. may apply filters and symmetry operators to the original model coordinates.
     * Return builder pointing to the new node. */
    assemblyStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'assembly',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.assembly),
            custom: params.custom,
            ref: params.ref,
        }));
    }
    /** Add a 'structure' node representing a "symmetry structure", i.e. applies symmetry operators to build crystal unit cells within given Miller indices.
     * Return builder pointing to the new node. */
    symmetryStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'symmetry',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry),
            custom: params.custom,
            ref: params.ref,
        }));
    }
    /** Add a 'structure' node representing a "symmetry mates structure", i.e. applies symmetry operators to build asymmetric units within a radius from the original model.
     * Return builder pointing to the new node. */
    symmetryMatesStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'symmetry_mates',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry_mates),
            custom: params.custom,
            ref: params.ref,
        }));
    }
    /** Add a 'volume' node representing raw volume data */
    volume(params = {}) {
        return new Volume(this._root, this.addChild('volume', params));
    }
    /** Add a 'coordinates' node indicating the parsed data type */
    coordinates(params = {}) {
        this.addChild('coordinates', params);
        return this;
    }
}
exports.Parse = Parse;
/** MVS builder pointing to a 'structure' node */
class Structure extends _Base {
    constructor() {
        super(...arguments);
        this.transform = bindMethod(this, TransformMixinImpl, 'transform');
        this.instance = bindMethod(this, TransformMixinImpl, 'instance');
        this.primitives = bindMethod(this, PrimitivesMixinImpl, 'primitives');
        this.primitivesFromUri = bindMethod(this, PrimitivesMixinImpl, 'primitivesFromUri');
    }
    /** Add a 'component' node and return builder pointing to it. 'component' node instructs to create a component (i.e. a subset of the parent structure). */
    component(params = {}) {
        var _a;
        const fullParams = { ...params, selector: (_a = params.selector) !== null && _a !== void 0 ? _a : 'all' };
        return new Component(this._root, this.addChild('component', fullParams));
    }
    /** Add a 'component_from_uri' node and return builder pointing to it. 'component_from_uri' node instructs to create a component defined by an external annotation resource. */
    componentFromUri(params) {
        return new Component(this._root, this.addChild('component_from_uri', params));
    }
    /** Add a 'component_from_source' node and return builder pointing to it. 'component_from_source' node instructs to create a component defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    componentFromSource(params) {
        return new Component(this._root, this.addChild('component_from_source', params));
    }
    /** Add a 'label_from_uri' node and return builder pointing back to the structure node. 'label_from_uri' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an external annotation resource. */
    labelFromUri(params) {
        this.addChild('label_from_uri', params);
        return this;
    }
    /** Add a 'label_from_source' node and return builder pointing back to the structure node. 'label_from_source' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    labelFromSource(params) {
        this.addChild('label_from_source', params);
        return this;
    }
    /** Add a 'tooltip_from_uri' node and return builder pointing back to the structure node. 'tooltip_from_uri' node instructs to add tooltips to parts of a structure. The tooltips are defined by an external annotation resource. */
    tooltipFromUri(params) {
        this.addChild('tooltip_from_uri', params);
        return this;
    }
    /** Add a 'tooltip_from_source' node and return builder pointing back to the structure node. 'tooltip_from_source' node instructs to add tooltips to parts of a structure. The tooltips are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    tooltipFromSource(params) {
        this.addChild('tooltip_from_source', params);
        return this;
    }
}
exports.Structure = Structure;
/** MVS builder pointing to a 'component' or 'component_from_uri' or 'component_from_source' node */
class Component extends _Base {
    constructor() {
        super(...arguments);
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
        this.transform = bindMethod(this, TransformMixinImpl, 'transform');
        this.instance = bindMethod(this, TransformMixinImpl, 'instance');
    }
    /** Add a 'representation' node and return builder pointing to it. 'representation' node instructs to create a visual representation of a component. */
    representation(params = {}) {
        var _a;
        const fullParams = { ...params, type: (_a = params.type) !== null && _a !== void 0 ? _a : 'cartoon' };
        return new Representation(this._root, this.addChild('representation', fullParams));
    }
    /** Add a 'label' node and return builder pointing back to the component node. 'label' node instructs to add a label (textual visual representation) to a component. */
    label(params) {
        this.addChild('label', params);
        return this;
    }
    /** Add a 'tooltip' node and return builder pointing back to the component node. 'tooltip' node instructs to add a text which is not a part of the visualization but should be presented to the users when they interact with the component (typically, the tooltip will be shown somewhere on the screen when the user hovers over a visual representation of the component). */
    tooltip(params) {
        this.addChild('tooltip', params);
        return this;
    }
}
exports.Component = Component;
/** MVS builder pointing to a 'representation' node */
class Representation extends _Base {
    /** Add a 'color' node and return builder pointing back to the representation node. 'color' node instructs to apply color to a visual representation. */
    color(params) {
        this.addChild('color', params);
        return this;
    }
    /** Add a 'color_from_uri' node and return builder pointing back to the representation node. 'color_from_uri' node instructs to apply colors to a visual representation. The colors are defined by an external annotation resource. */
    colorFromUri(params) {
        this.addChild('color_from_uri', params);
        return this;
    }
    /** Add a 'color_from_source' node and return builder pointing back to the representation node. 'color_from_source' node instructs to apply colors to a visual representation. The colors are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    colorFromSource(params) {
        this.addChild('color_from_source', params);
        return this;
    }
    /** Add an 'opacity' node and return builder pointing back to the representation node. 'opacity' node instructs to customize opacity/transparency of a visual representation. */
    opacity(params) {
        this.addChild('opacity', params);
        return this;
    }
    /** Add a 'clip' node and return builder pointing back to the representation node. 'clip' node instructs to apply clipping to a visual representation. */
    clip(params) {
        this.addChild('clip', params);
        return this;
    }
}
exports.Representation = Representation;
/** MVS builder pointing to a 'component' or 'component_from_uri' or 'component_from_source' node */
class Volume extends _Base {
    constructor() {
        super(...arguments);
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
        this.transform = bindMethod(this, TransformMixinImpl, 'transform');
        this.instance = bindMethod(this, TransformMixinImpl, 'instance');
    }
    /** Add a 'representation' node and return builder pointing to it. 'representation' node instructs to create a visual representation of a component. */
    representation(params) {
        if (!params) {
            params = { type: 'isosurface' };
        }
        return new VolumeRepresentation(this._root, this.addChild('volume_representation', params));
    }
}
exports.Volume = Volume;
/** MVS builder pointing to a 'volume_representation' node */
class VolumeRepresentation extends _Base {
    constructor() {
        super(...arguments);
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
    }
    /** Add a 'color' node and return builder pointing back to the representation node. 'color' node instructs to apply color to a visual representation. */
    color(params) {
        this.addChild('color', params);
        return this;
    }
    /** Add an 'opacity' node and return builder pointing back to the representation node. 'opacity' node instructs to customize opacity/transparency of a visual representation. */
    opacity(params) {
        this.addChild('opacity', params);
        return this;
    }
    /** Add a 'clip' node and return builder pointing back to the representation node. 'clip' node instructs to apply clipping to a visual representation. */
    clip(params) {
        this.addChild('clip', params);
        return this;
    }
}
exports.VolumeRepresentation = VolumeRepresentation;
/** MVS builder pointing to a 'primitives' node */
class Primitives extends _Base {
    constructor() {
        super(...arguments);
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
    }
    /** Construct custom meshes/shapes in a low-level fashion by providing vertices and indices. */
    mesh(params) {
        this.addChild('primitive', { kind: 'mesh', ...params });
        return this;
    }
    /** Construct custom set of lines in a low-level fashion by providing vertices and indices. */
    lines(params) {
        this.addChild('primitive', { kind: 'lines', ...params });
        return this;
    }
    /** Defines a tube (3D cylinder), connecting a start and an end point. */
    tube(params) {
        this.addChild('primitive', { kind: 'tube', ...params });
        return this;
    }
    /** Defines an arrow. */
    arrow(params) {
        this.addChild('primitive', { kind: 'arrow', ...params });
        return this;
    }
    /** Defines a tube, connecting a start and an end point, with label containing distance between start and end. */
    distance(params) {
        this.addChild('primitive', { kind: 'distance_measurement', ...params });
        return this;
    }
    /** Defines an angle between vectors (b - a) and (c - b). */
    angle(params) {
        this.addChild('primitive', { kind: 'angle_measurement', ...params });
        return this;
    }
    /** Defines a label. */
    label(params) {
        this.addChild('primitive', { kind: 'label', ...params });
        return this;
    }
    /** Defines an ellipse. */
    ellipse(params) {
        this.addChild('primitive', { kind: 'ellipse', ...params });
        return this;
    }
    /** Defines an ellipsoid. */
    ellipsoid(params) {
        this.addChild('primitive', { kind: 'ellipsoid', ...params });
        return this;
    }
    /** Defines a sphere (a special case of ellipsoid). */
    sphere(params) {
        this.addChild('primitive', { kind: 'ellipsoid', ...params });
        return this;
    }
    /** Defines a box. */
    box(params) {
        this.addChild('primitive', { kind: 'box', ...params });
        return this;
    }
    /** Add a 'clip' node and return builder pointing back to the representation node. 'clip' node instructs to apply clipping to a visual representation. */
    clip(params) {
        this.addChild('clip', params);
        return this;
    }
}
exports.Primitives = Primitives;
/** MVS builder pointing to a 'primitives_from_uri' node */
class PrimitivesFromUri extends _Base {
    constructor() {
        super(...arguments);
        this.focus = bindMethod(this, FocusMixinImpl, 'focus');
    }
    /** Add a 'clip' node and return builder pointing back to the representation node. 'clip' node instructs to apply clipping to a visual representation. */
    clip(params) {
        this.addChild('clip', params);
        return this;
    }
}
function bindMethod(thisObj, mixin, methodName) {
    return mixin.prototype[methodName].bind(thisObj);
}
class FocusMixinImpl extends _Base {
    focus(params = {}) {
        this.addChild('focus', params);
        return this;
    }
}
;
;
class PrimitivesMixinImpl extends _Base {
    primitives(params = {}) {
        return new Primitives(this._root, this.addChild('primitives', params));
    }
    primitivesFromUri(params) {
        return new PrimitivesFromUri(this._root, this.addChild('primitives_from_uri', params));
    }
}
;
;
class TransformMixinImpl extends _Base {
    transform(params = {}) {
        validateTransformParams(params);
        this.addChild('transform', params);
        return this;
    }
    instance(params = {}) {
        validateTransformParams(params);
        this.addChild('instance', params);
        return this;
    }
}
;
function validateTransformParams(params) {
    if (params.rotation && params.rotation.length !== 9) {
        throw new Error('ValueError: `rotation` parameter must be an array of 9 numbers');
    }
    if (params.matrix && params.matrix.length !== 16) {
        throw new Error('ValueError: `matrix` parameter must be an array of 16 numbers');
    }
    if (params.matrix && (params.translation || params.rotation)) {
        throw new Error('ValueError: `matrix` parameter cannot be used together with `translation` or `rotation` parameters');
    }
}
/** Demonstration of usage of MVS builder */
function builderDemo() {
    const builder = createMVSBuilder();
    builder.canvas({ background_color: 'white' });
    const struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
    struct.component().representation().color({ color: 'white' });
    struct.component({ selector: 'ligand' }).representation({ type: 'ball_and_stick', custom: { repr_quality: 'high' }, ref: 'Ligand' })
        .color({ color: '#555555' })
        .color({ selector: { type_symbol: 'N' }, color: '#3050F8' })
        .color({ selector: { type_symbol: 'O' }, color: '#FF0D0D' })
        .color({ selector: { type_symbol: 'S' }, color: '#FFFF30' })
        .color({ selector: { type_symbol: 'FE' }, color: '#E06633' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '1' }).component().representation().color({ color: 'cyan' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '2' }).component().representation().color({ color: 'blue' });
    const cif = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1wrf_updated.cif' }).parse({ format: 'mmcif' });
    cif.modelStructure({ model_index: 0 }).component().representation().color({ color: '#CC0000' });
    cif.modelStructure({ model_index: 1 }).component().representation().color({ color: '#EE7700' });
    cif.modelStructure({ model_index: 2 }).component().representation().color({ color: '#FFFF00' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [30, 0, 0] }).component().representation().color({ color: '#ff88bb' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [60, 0, 0], rotation: [0, 1, 0, -1, 0, 0, 0, 0, 1] }).component().representation().color({ color: '#aa0077' });
    return builder.getState();
}
;
function splitParams(params_custom_ref) {
    const { custom, ref, ...params } = params_custom_ref;
    return { params: params, custom, ref };
}
