// molstar/lib/commonjs/mol-repr/structure/visual/gaussian-density-volume.js
"use strict";
/**
 * Copyright (c) 2018-2025 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitsGaussianDensityVolumeParams = exports.GaussianDensityVolumeParams = void 0;
exports.GaussianDensityVolumeVisual = GaussianDensityVolumeVisual;
exports.UnitsGaussianDensityVolumeVisual = UnitsGaussianDensityVolumeVisual;
const param_definition_1 = require("../../../mol-util/param-definition.js");
const gaussian_1 = require("./util/gaussian.js");
const direct_volume_1 = require("../../../mol-geo/geometry/direct-volume/direct-volume.js");
const complex_visual_1 = require("../complex-visual.js");
const linear_algebra_1 = require("../../../mol-math/linear-algebra.js");
const element_1 = require("./util/element.js");
const geometry_1 = require("../../../mol-math/geometry.js");
const units_visual_1 = require("../units-visual.js");
function createGaussianDensityVolume(ctx, structure, theme, props, directVolume) {
    const { webgl } = ctx;
    if (!webgl) {
        // gpu gaussian density also needs blendMinMax but there is no fallback here so
        // we allow it here with the results that there is no group id assignment and
        // hence no group-based coloring or picking
        throw new Error('GaussianDensityVolume requires `webgl`');
    }
    const axisOrder = linear_algebra_1.Vec3.create(0, 1, 2);
    const stats = { min: 0, max: 1, mean: 0.04, sigma: 0.01 };
    const create = (directVolume) => {
        const oldTexture = directVolume ? directVolume.gridTexture.ref.value : undefined;
        const densityTextureData = (0, gaussian_1.computeStructureGaussianDensityTexture)(structure, theme.size, props, webgl, oldTexture);
        const { transform, texture, bbox, gridDim } = densityTextureData;
        const unitToCartn = linear_algebra_1.Mat4.mul((0, linear_algebra_1.Mat4)(), transform, linear_algebra_1.Mat4.fromScaling((0, linear_algebra_1.Mat4)(), gridDim));
        const cellDim = linear_algebra_1.Mat4.getScaling((0, linear_algebra_1.Vec3)(), transform);
        const vol = direct_volume_1.DirectVolume.create(bbox, gridDim, transform, unitToCartn, cellDim, texture, stats, true, axisOrder, 'byte', directVolume);
        const sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), structure.boundary.sphere, densityTextureData.maxRadius);
        vol.setBoundingSphere(sphere);
        return vol;
    };
    const vol = create(directVolume);
    vol.meta.reset = () => {
        create(vol);
    };
    return vol;
}
exports.GaussianDensityVolumeParams = {
    ...complex_visual_1.ComplexDirectVolumeParams,
    ...gaussian_1.GaussianDensityParams,
    ignoreHydrogens: param_definition_1.ParamDefinition.Boolean(false),
    ignoreHydrogensVariant: param_definition_1.ParamDefinition.Select('all', param_definition_1.ParamDefinition.arrayToOptions(['all', 'non-polar'])),
    includeParent: param_definition_1.ParamDefinition.Boolean(false, { isHidden: true }),
};
function GaussianDensityVolumeVisual(materialId) {
    return (0, complex_visual_1.ComplexDirectVolumeVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.GaussianDensityVolumeParams),
        createGeometry: createGaussianDensityVolume,
        createLocationIterator: element_1.ElementIterator.fromStructure,
        getLoci: element_1.getSerialElementLoci,
        eachLocation: element_1.eachSerialElement,
        setUpdateState: (state, newProps, currentProps) => {
            if (newProps.resolution !== currentProps.resolution)
                state.createGeometry = true;
            if (newProps.radiusOffset !== currentProps.radiusOffset)
                state.createGeometry = true;
            if (newProps.smoothness !== currentProps.smoothness)
                state.createGeometry = true;
            if (newProps.ignoreHydrogens !== currentProps.ignoreHydrogens)
                state.createGeometry = true;
            if (newProps.ignoreHydrogensVariant !== currentProps.ignoreHydrogensVariant)
                state.createGeometry = true;
            if (newProps.traceOnly !== currentProps.traceOnly)
                state.createGeometry = true;
            if (newProps.includeParent !== currentProps.includeParent)
                state.createGeometry = true;
        },
        dispose: (geometry) => {
            geometry.gridTexture.ref.value.destroy();
        }
    }, materialId);
}
//
function createUnitsGaussianDensityVolume(ctx, unit, structure, theme, props, directVolume) {
    const { webgl } = ctx;
    if (!webgl) {
        // gpu gaussian density also needs blendMinMax but there is no fallback here so
        // we allow it here with the results that there is no group id assignment and
        // hence no group-based coloring or picking
        throw new Error('GaussianDensityVolume requires `webgl`');
    }
    const axisOrder = linear_algebra_1.Vec3.create(0, 1, 2);
    const stats = { min: 0, max: 1, mean: 0.04, sigma: 0.01 };
    const create = (directVolume) => {
        const oldTexture = directVolume ? directVolume.gridTexture.ref.value : undefined;
        const densityTextureData = (0, gaussian_1.computeUnitGaussianDensityTexture)(structure, unit, theme.size, props, webgl, oldTexture);
        const { transform, texture, bbox, gridDim } = densityTextureData;
        const unitToCartn = linear_algebra_1.Mat4.mul((0, linear_algebra_1.Mat4)(), transform, linear_algebra_1.Mat4.fromScaling((0, linear_algebra_1.Mat4)(), gridDim));
        const cellDim = linear_algebra_1.Mat4.getScaling((0, linear_algebra_1.Vec3)(), transform);
        const vol = direct_volume_1.DirectVolume.create(bbox, gridDim, transform, unitToCartn, cellDim, texture, stats, true, axisOrder, 'byte', directVolume);
        const sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, densityTextureData.maxRadius);
        vol.setBoundingSphere(sphere);
        return vol;
    };
    const vol = create(directVolume);
    vol.meta.reset = () => {
        create(vol);
    };
    return vol;
}
exports.UnitsGaussianDensityVolumeParams = {
    ...units_visual_1.UnitsDirectVolumeParams,
    ...gaussian_1.GaussianDensityParams,
    ignoreHydrogens: param_definition_1.ParamDefinition.Boolean(false),
    ignoreHydrogensVariant: param_definition_1.ParamDefinition.Select('all', param_definition_1.ParamDefinition.arrayToOptions(['all', 'non-polar'])),
    includeParent: param_definition_1.ParamDefinition.Boolean(false, { isHidden: true }),
};
function UnitsGaussianDensityVolumeVisual(materialId) {
    return (0, units_visual_1.UnitsDirectVolumeVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.UnitsGaussianDensityVolumeParams),
        createGeometry: createUnitsGaussianDensityVolume,
        createLocationIterator: element_1.ElementIterator.fromGroup,
        getLoci: element_1.getElementLoci,
        eachLocation: element_1.eachElement,
        setUpdateState: (state, newProps, currentProps) => {
            if (newProps.resolution !== currentProps.resolution)
                state.createGeometry = true;
            if (newProps.radiusOffset !== currentProps.radiusOffset)
                state.createGeometry = true;
            if (newProps.smoothness !== currentProps.smoothness)
                state.createGeometry = true;
            if (newProps.ignoreHydrogens !== currentProps.ignoreHydrogens)
                state.createGeometry = true;
            if (newProps.ignoreHydrogensVariant !== currentProps.ignoreHydrogensVariant)
                state.createGeometry = true;
            if (newProps.traceOnly !== currentProps.traceOnly)
                state.createGeometry = true;
            if (newProps.includeParent !== currentProps.includeParent)
                state.createGeometry = true;
        },
        dispose: (geometry) => {
            geometry.gridTexture.ref.value.destroy();
        }
    }, materialId);
}
