// molstar/lib/commonjs/mol-model-props/computed/interactions/hydrophobic.js
"use strict";
/**
 * Copyright (c) 2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author Fred Ludlow <Fred.Ludlow@astx.com>
 *
 * based in part on NGL (https://github.com/arose/ngl)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HydrophobicProvider = exports.HydrophobicAtomProvider = void 0;
const param_definition_1 = require("../../../mol-util/param-definition.js");
const features_1 = require("./features.js");
const util_1 = require("../chemistry/util.js");
const types_1 = require("../../../mol-model/structure/model/properties/atomic/types.js");
const common_1 = require("./common.js");
const HydrophobicParams = {
    distanceMax: param_definition_1.ParamDefinition.Numeric(4.0, { min: 1, max: 5, step: 0.1 }),
};
/**
 * Hydropbobic atoms
 * - Carbon only bonded to carbon or hydrogen
 * - Fluorine
 */
function addHydrophobicAtom(structure, unit, builder) {
    const { elements } = unit;
    const { x, y, z } = unit.model.atomicConformation;
    for (let i = 0, il = elements.length; i < il; ++i) {
        const element = (0, util_1.typeSymbol)(unit, i);
        let flag = false;
        if (element === types_1.Elements.C) {
            flag = true;
            (0, util_1.eachBondedAtom)(structure, unit, i, (unitB, indexB) => {
                const elementB = (0, util_1.typeSymbol)(unitB, indexB);
                if (elementB !== types_1.Elements.C && elementB !== types_1.Elements.H)
                    flag = false;
            });
        }
        else if (element === types_1.Elements.F) {
            flag = true;
        }
        if (flag) {
            builder.add(common_1.FeatureType.HydrophobicAtom, common_1.FeatureGroup.None, x[elements[i]], y[elements[i]], z[elements[i]], i);
        }
    }
}
function isHydrophobicContact(ti, tj) {
    return ti === common_1.FeatureType.HydrophobicAtom && tj === common_1.FeatureType.HydrophobicAtom;
}
function testHydrophobic(structure, infoA, infoB, distanceSq) {
    const typeA = infoA.types[infoA.feature];
    const typeB = infoB.types[infoB.feature];
    if (!isHydrophobicContact(typeA, typeB))
        return;
    const indexA = infoA.members[infoA.offsets[infoA.feature]];
    const indexB = infoB.members[infoB.offsets[infoB.feature]];
    if ((0, util_1.typeSymbol)(infoA.unit, indexA) === types_1.Elements.F && (0, util_1.typeSymbol)(infoB.unit, indexB) === types_1.Elements.F)
        return;
    return common_1.InteractionType.Hydrophobic;
}
//
exports.HydrophobicAtomProvider = features_1.Features.Provider([common_1.FeatureType.HydrophobicAtom], addHydrophobicAtom);
exports.HydrophobicProvider = {
    name: 'hydrophobic',
    params: HydrophobicParams,
    createTester: (props) => {
        return {
            maxDistance: props.distanceMax,
            requiredFeatures: new Set([common_1.FeatureType.HydrophobicAtom]),
            getType: (structure, infoA, infoB, distanceSq) => testHydrophobic(structure, infoA, infoB, distanceSq)
        };
    }
};
