// molstar/lib/commonjs/mol-repr/structure/params.js
"use strict";
/**
 * Copyright (c) 2018-2025 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureImageParams = exports.StructureTextureMeshParams = exports.StructureDirectVolumeParams = exports.StructureTextParams = exports.StructureLinesParams = exports.StructurePointsParams = exports.StructureCylindersParams = exports.StructureSpheresParams = exports.StructureMeshParams = exports.StructureParams = void 0;
exports.getUnitKindsParam = getUnitKindsParam;
const direct_volume_1 = require("../../mol-geo/geometry/direct-volume/direct-volume.js");
const lines_1 = require("../../mol-geo/geometry/lines/lines.js");
const mesh_1 = require("../../mol-geo/geometry/mesh/mesh.js");
const points_1 = require("../../mol-geo/geometry/points/points.js");
const spheres_1 = require("../../mol-geo/geometry/spheres/spheres.js");
const cylinders_1 = require("../../mol-geo/geometry/cylinders/cylinders.js");
const text_1 = require("../../mol-geo/geometry/text/text.js");
const texture_mesh_1 = require("../../mol-geo/geometry/texture-mesh/texture-mesh.js");
const image_1 = require("../../mol-geo/geometry/image/image.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const common_1 = require("./visual/util/common.js");
function getUnitKindsParam(defaultValue) {
    return param_definition_1.ParamDefinition.MultiSelect(defaultValue, common_1.UnitKindOptions, { description: 'For which kinds of units/chains to show the representation visuals.' });
}
exports.StructureParams = {
    unitKinds: getUnitKindsParam(['atomic', 'spheres']),
    includeParent: param_definition_1.ParamDefinition.Boolean(false, { isHidden: true }),
};
exports.StructureMeshParams = { ...mesh_1.Mesh.Params };
exports.StructureSpheresParams = { ...spheres_1.Spheres.Params };
exports.StructureCylindersParams = { ...cylinders_1.Cylinders.Params };
exports.StructurePointsParams = { ...points_1.Points.Params };
exports.StructureLinesParams = { ...lines_1.Lines.Params };
exports.StructureTextParams = { ...text_1.Text.Params };
exports.StructureDirectVolumeParams = { ...direct_volume_1.DirectVolume.Params };
exports.StructureTextureMeshParams = { ...texture_mesh_1.TextureMesh.Params };
exports.StructureImageParams = { ...image_1.Image.Params };
